GC fixes and doc

muellmusik opened the pull request:
This is an improved version of #1615. In addition to fixes there (updated after comments, and whitespace corrected) it renames the ```collect``` arg in object creation funcs to ```runGC``` in order to make its purpose clearer (this was a bit confusing), and substantially updates the Writing Primitives helpfile (understanding is the first step in not hosing the GC...).

muellmusik commented:
@timblechmann Just poking around with this, I have another interesting bug. On this branch run the following:
```
Array.fill(10000, {1.0.rand}).unlace(1000) // do this twice
```
On the second time, I reliably get:
```
grey count error
caught exception in primitive Array:prUnlace
ERROR: grey count error
grey count error
caught exception in primitive Object:asString
ERROR: grey count error
grey count error
caught exception in runLibrary Collection:collectAs -378
BYTECODES: (15)
  0   32		 PushTempZeroVar 'class'
  1   C0 09    SendSpecialMsg 'size'
  3   C2 00    SendSpecialMsg 'new'
  5   80 03    StoreTempVar 'res'
  7   60       PushSpecialValue this
  8   04 00    PushLiteralX instance of FunctionDef in Method Collection:collectAs
 10   C2 1D    SendSpecialMsg 'do'
 12   F0       Drop
 13   33		 PushTempZeroVar 'res'
 14   F3       Return
ERROR: grey count error
```
I do not think this is caused by b03b834 (I'm not sure how it could be anyway), as if you run the same example on master, you can also get a grey count error, just less reliably and without the bytecodes.

973314d is a possibility for introducing the issue, and indeed, if I remove ```SetObject(b, obj2);``` and set the inner object instantiations not to collect, it seems to solve the problem. But I am not sure why this happens (it looks correct) and it would be good to understand the issue to ensure that things are working properly. FWIW I tried to make ```allTuples``` fail (it uses the same approach) but couldn't.

One clue perhaps is that doing it this way seems not to cause the problem:
```
// execute these lines one at a time repeatedly
x = Array.fill(10000, {1.0.rand});
x.unlace(1000);
```
Could you take a look and see if you have any ideas?

timblechmann commented:
@muellmusik checking `prArrayUnlace`: it seems `obj2->size = numLists` sets the object size, but the gc runs before all objects are initialized. being lucky the gc does not find dangling pointers.

muellmusik commented:
So the correct fix is to increment obj2's size in the for loop, I assume. (Which is why prArrayAllTuples doesn't have this problem.) I'll add another commit for this, and run a quick review on all size settings.

muellmusik commented:
Okay, I have fixed this in ```prArrayUnlace``` and fixed two similar issues, in ```prArrayMultiChannelExpand``` and ```prLanguageConfig_getLibraryPaths```. As well, I've removed some unnecessary initialisations, one unnecessary call to GCWrite, and improved the doc a little.

Please review carefully. If approved, these fixes should definitely be in 3.8.

muellmusik commented:
Sorry, meant 3.7!

muellmusik commented:
Okay, since this is now a little messy again, I'll submit another clean version in a bit.
